File: Web_controleur/moteur_control/inverse_kinematics.py
import math
import matplotlib.pyplot as plt

def inverse_kinematics_2d_flexible(x, y, d1, d2, theta_max_deg):
    r = math.hypot(x, y)

    max_reach = d1 + d2
    min_reach = abs(d1 - d2)

    # Conversion de l'angle maximum en radians
    theta_max = math.radians(theta_max_deg)

    # Clamp la distance si hors de portée
    if r > max_reach:
        x *= max_reach / r
        y *= max_reach / r
        r = max_reach
    elif r < min_reach:
        x *= min_reach / r
        y *= min_reach / r
        r = min_reach

    # Angle du genou (theta)
    cos_theta = (r ** 2 - d1 ** 2 - d2 ** 2) / (2 * d1 * d2)
    cos_theta = max(-1, min(1, cos_theta))
    theta = math.acos(cos_theta)

    # Limitation du genou
    if theta > theta_max:
        theta = theta_max  # on le bride

        # Nouvelle distance atteignable avec ce theta_max
        effective_r = math.sqrt(d1 ** 2 + d2 ** 2 + 2 * d1 * d2 * math.cos(theta))

        x *= effective_r / r
        y *= effective_r / r
        r = effective_r

    # Angle vers la cible
    alpha = math.atan2(y, x)

    # Calcul de beta
    cos_beta = (r ** 2 + d2 ** 2 - d1 ** 2) / (2 * r * d2)
    cos_beta = max(-1, min(1, cos_beta))
    beta = math.acos(cos_beta)

    # Angle du 1er segment
    phi = alpha - beta

    return phi, theta


def draw_leg(x_target, y_target, d1, d2):
    theta_max_deg = 150

    phi, theta = inverse_kinematics_2d_flexible(x_target, y_target, d1, d2, theta_max_deg)

    #phi = 0
    #theta = 0



    # Position genou
    x1 = d2 * math.cos(phi)
    y1 = d2 * math.sin(phi)

    # Position pied
    x2 = x1 + d1 * math.cos(phi + theta)
    y2 = y1 + d1 * math.sin(phi + theta)

    # Affichage des angles
    print(f"phi  = {phi:.3f} rad  ({math.degrees(phi):.2f}°)")
    print(f"theta = {theta:.3f} rad  ({math.degrees(theta):.2f}°)")

    # Tracé
    plt.figure(figsize=(6, 6))
    plt.plot([0, x1], [0, y1], 'ro-', label="Segment 1 (d2)")
    plt.plot([x1, x2], [y1, y2], 'bo-', label="Segment 2 (d1)")
    plt.plot(x_target, y_target, 'gx', markersize=10, label="Cible")
    plt.title(f"Patte articulée avec θ max = {theta_max_deg}°")
    plt.xlabel("x")
    plt.ylabel("y")
    plt.axis('equal')
    plt.grid(True)
    plt.legend()
    plt.xlim(-d1 - d2 - 1, d1 + d2 + 1)
    plt.ylim(-d1 - d2 - 1, d1 + d2 + 1)
    plt.show()


# Exemple d’utilisation :
draw_leg(x_target=0, y_target=1, d1=4, d2=4)

File: Microcontroleur_vesc_comm/src/InverseKinematics.cpp
#include "InverseKinematics.h"
#include <math.h>

AngleMoteur inverse_kinematics_2d_flexible(float x, float y, float d1, float d2, float theta_max_deg) {
  float r = hypot(x, y);

  float max_reach = d1 + d2;
  float min_reach = fabs(d1 - d2);

  // Conversion en radians pour le calcul
  float theta_max = theta_max_deg * M_PI / 180.0f;

  // Clamp de la distance
  if (r > max_reach) {
    x *= max_reach / r;
    y *= max_reach / r;
    r = max_reach;
  } else if (r < min_reach && r > 1e-6f) { // éviter div/0
    x *= min_reach / r;
    y *= min_reach / r;
    r = min_reach;
  }

  // Angle du genou (theta en radian)
  float cos_theta = (r * r - d1 * d1 - d2 * d2) / (2.0f * d1 * d2);
  if (cos_theta > 1.0f) cos_theta = 1.0f;
  if (cos_theta < -1.0f) cos_theta = -1.0f;
  float theta = acos(cos_theta);

  // Limitation du genou
  if (theta > theta_max) {
    theta = theta_max;

    float effective_r = sqrt(d1 * d1 + d2 * d2 + 2.0f * d1 * d2 * cos(theta));
    if (r > 1e-6f) {
      x *= effective_r / r;
      y *= effective_r / r;
    }
    r = effective_r;
  }

  // Angle vers la cible
  float alpha = atan2(y, x);

  // Calcul de beta
  float cos_beta = (r * r + d2 * d2 - d1 * d1) / (2.0f * r * d2);
  if (cos_beta > 1.0f) cos_beta = 1.0f;
  if (cos_beta < -1.0f) cos_beta = -1.0f;
  float beta = acos(cos_beta);

  // Angle du 1er segment (radians → degrés)
  float phi = (alpha - beta) * 180.0f / M_PI;
  theta = theta * 180.0f / M_PI;

  AngleMoteur result = {phi, theta};
  return result;
}

MaxAngleMoteur compute_max_angle(AngleMoteur current, AngleMoteur target, float max_angle) {
    MaxAngleMoteur result;
    
    float diff1 = target.phi - current.phi;
    float diff2 = target.theta - current.theta;

    if (fabs(diff1) > fabs(diff2)) {
        result.maxAngle_phi = max_angle;
        result.maxAngle_theta = fabs(diff2 / diff1) * max_angle;
    } else {
        result.maxAngle_theta = max_angle;
        result.maxAngle_phi = fabs(diff1 / diff2) * max_angle;
    }

    return result;
}

DataFunction interpolate_data_point(DataPoint (*func)(float), float a,float d1, float d2, float theta_max_deg, AngleMoteur actualAngles) {
    DataPoint p1 = func(a);
    DataFunction result;
    
    AngleMoteur angles = inverse_kinematics_2d_flexible(p1.x, p1.y, d1, d2, theta_max_deg);
    MaxAngleMoteur maxAngles = compute_max_angle(actualAngles, angles, p1.maxAngle);

    result.maxAngle_phi = maxAngles.maxAngle_phi;
    result.maxAngle_theta = maxAngles.maxAngle_theta;
    result.phi = angles.phi;
    result.theta = angles.theta;


    return result;
}